src/MarkedOptions.ts

```diff
@@ -169,46 +173,45 @@ export interface MarkedExtension {
    * Each token is passed by reference so updates are persisted when passed to the parser.
    * The return value of the function is ignored.
    */
-  walkTokens?: ((token: Token) => void | Promise<void>) | undefined | null;
+  walkTokens?: ((token: Token) => void | unknown | Promise<void>) | undefined | null;
   /**
    * Generate closing slash for self-closing tags (<br/> instead of <br>)
    * @deprecated Deprecated in v5.0.0 use marked-xhtml to emit self-closing HTML tags for void elements (<br/>, <img/>, etc.) with a "/" as required by XHTML.
    */
   xhtml?: boolean | undefined;
 }
 
-export interface MarkedOptions extends Omit<MarkedExtension, 'extensions' | 'renderer' | 'tokenizer' | 'walkTokens'> {
+export interface MarkedOptions extends Omit<MarkedExtension, 'renderer' | 'tokenizer' | 'extensions' | 'walkTokens'> {
   /**
    * Type: object Default: new Renderer()
    *
    * An object containing functions to render tokens to HTML.
    */
-  renderer?: Omit<_Renderer, 'constructor'> | undefined | null;
+  renderer?: _Renderer | undefined | null;
 
   /**
    * The tokenizer defines how to turn markdown text into tokens.
    */
-  tokenizer?: Omit<_Tokenizer, 'constructor'> | undefined | null;
+  tokenizer?: _Tokenizer | undefined | null;
 
   /**
-   * The walkTokens function gets called with every token.
-   * Child tokens are called before moving on to sibling tokens.
-   * Each token is passed by reference so updates are persisted when passed to the parser.
-   * The return value of the function is ignored.
+   * Custom extensions
    */
-  walkTokens?: ((token: Token) => void | Promise<void> | Array<void | Promise<void>>) | undefined | null;
+  extensions?: null | {
+    renderers: {
+      [name: string]: RendererExtensionFunction;
+    };
+    childTokens: {
+      [name: string]: string[];
+    };
+    inline?: TokenizerExtensionFunction[];
+    block?: TokenizerExtensionFunction[];
+    startInline?: TokenizerStartFunction[];
+    startBlock?: TokenizerStartFunction[];
+  };
 
   /**
-   * Add tokenizers and renderers to marked
+   * walkTokens function returns array of values for Promise.all
    */
-  extensions?:
-    | (TokenizerAndRendererExtension[] & {
-    renderers: Record<string, RendererExtensionFunction>,
-    childTokens: Record<string, string[]>,
-    block: any[],
-    inline: any[],
-    startBlock: Array<(this: TokenizerThis, src: string) => number | void>,
-    startInline: Array<(this: TokenizerThis, src: string) => number | void>
-  })
-    | undefined | null;
+  walkTokens?: null | ((token: Token) => void | (unknown | Promise<void>)[]);
```

Comment:
This breaks downstream compilation because `(unknown | Promise<void>)[]` gets collapsed to `unknown[]` https://github.com/jupyterlab/jupyterlab/issues/15020 see https://stackoverflow.com/questions/61685819/typescript-does-not-recognize-a-union-of-type-unknown-and-type-promiseunknown

Should we adapt the type donwstream, or do you consider it a bug?

Reply:
@UziTech Perhaps this should be `(void | Promise<void>)[]` instead?

Reply:
Ya I think `(void | Promise<void>)[]` should work

Reply:
The only time we care about the return value is if `async` is true. Than we await the return value. I think the problem was that we can't await `void` in typescript but it is perfectly fine in JavaScript.

Reply:
> The only time we care about the return value is if async is true.

Maybe we need different types when `async: true` if we want it to be more precise

Reply:
Should be fixed by #2955